src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -4,92 +4,91 @@
-                c.put(REGION_PREFIX + language, cacheItem);
-            }
+    public Map<String, String> getRegionsFromCLDR(String locale){
+        locale = locale.replace("_", "-").toLowerCase();
+        Map<String, String> regionMap = null;
+        logger.debug("Look for region list from cache for locale [{}]", locale);
+        FormattingCacheService formattingCacheService = new FormattingCacheService();
+        regionMap = formattingCacheService.getRegions(locale);
+        if (regionMap != null) {
+            logger.debug("Find regions from cache for locale [{}]!", locale);
+            return regionMap;
         }
-        return respMap;
+        regionMap = getRegionsFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        if (regionMap != null) {
+            logger.debug("Find the regions for locale [{}].\n", locale);
+            formattingCacheService.addRegions(locale, regionMap);
+            logger.debug("Regions is cached for locale [{}]!\n\n", locale);
+            return regionMap;
+        }
+        return regionMap;
     }
 
-    public Map<String, String> getDisplayNamesFromCLDR(String language, 
-    		ListIterator<DataSourceEnum> msgSourceQueueIter) {
-    	Map<String, String> dispMap = new HashMap<String, String>(); 	
-    	if (!msgSourceQueueIter.hasNext()) 
-    		return dispMap;
-        
-        logger.trace("look for displayNames from cache");
-        Cache c = VIPCfg.getInstance().getCacheManager()
-                .getCache(VIPCfg.CACHE_L2);
-        if (c != null) {
-        	FormatCacheItem cacheItem = (FormatCacheItem) c.get(DISPN_PREFIX + language); 
-        	if (cacheItem == null) {
-        		cacheItem = new FormatCacheItem();
-        	}
-            dispMap = cacheItem.getCachedData();
-            if (dispMap.isEmpty()) {
-            	DataSourceEnum dataSource = (DataSourceEnum) msgSourceQueueIter.next();
-            	Map<String, String> tmpMap = dataSource.createLocaleOpt().getLanguages(language);
-                dispMap = JSONUtils.map2SortMap(tmpMap);
-                if (dispMap != null && dispMap.size() > 0) {
-                    c.put(DISPN_PREFIX + language, new FormatCacheItem(dispMap));
-                }
-            }
+    private Map<String, String> getRegionsFromDS(String locale, ListIterator<DataSourceEnum> msgSourceQueueIter) {
+        Map<String, String> regions = null;
+        if (!msgSourceQueueIter.hasNext())
+            return regions;
+        DataSourceEnum dataSource = (DataSourceEnum) msgSourceQueueIter.next();
+        regions = dataSource.createLocaleOpt(dto).getRegions(locale);
+        if (regions == null || regions.isEmpty()) {
+            regions = getRegionsFromDS(locale, msgSourceQueueIter);
         }
-        
-        if (dispMap == null || dispMap.isEmpty()) {
-        	return getDisplayNamesFromCLDR(language, msgSourceQueueIter);
+        return regions;
+    }
```

Comment:
As we discussed offline, add a new class LocaleDTO in case parameter increase, but won’t put locale into it.
Because 'locale' is a frequently changed parameter, put it into DTO will need new service instance multiple times when its value changed, and new instance is time consuming.